### ISSUE TITLE
Unimplemented: system segment descriptor

### ISSUE THREAD

**Issue by perlun** [copy/v86#157]
First, thanks for an admirable piece of software! I'm currently looking into getting our own homebrewed OS running in v86 (https://github.com/chaos4ever/chaos/issues/80), but ran into an issue when booting it up:

![image](https://cloud.githubusercontent.com/assets/630613/26029597/42989f32-3841-11e7-9515-6e80f10e78c5.png)

A textual version of the error comes here:

```
Unimplemented: load system segment descriptor, type = 9 (Available 386 TSS)  v86_all.js:405:721
uncaught exception: Unimplemented: load system segment descriptor, type = 9 (Available 386 TSS)  (unknown)
Execution stopped  v86_all.js:405:721
Unimplemented: load system segment descriptor, type = 9 (Available 386 TSS)  v86_all.js:179:264
```

I am a part of the core team that wrote this OS back in the days, but have regretfully not been doing so much low-level development lately, so I barely remember this. What is a "system segment descriptor", type 9? 😄 I guess I can start digging into our codebase to see how we use it, but... is our kernel a goofball in that it uses weird stuff that no other normal system use, or what's the deal? Any suggestions for how to refactor our code to not use system segment descriptors? (yes, I know it's incredibly hard for someone who hasn't studied our internals to answer that... but if you have ideas, I'm very open for them.)

Once again, a big **thanks** for all your effort with this!

**Comment by copy** [comment]
> I guess I can start digging into our codebase to see how we use it, but... is our kernel a goofball in that it uses weird stuff that no other normal system use, or what's the deal?

Yes, this is the case, although I've come across several other kernels that use this feature. This feature is hardware multitasking, see https://pdos.csail.mit.edu/6.828/2011/readings/i386/c07.htm

> Any suggestions for how to refactor our code to not use system segment descriptors?

For starters, the steps described in https://pdos.csail.mit.edu/6.828/2011/readings/i386/s07_05.htm need to be executed manually and the call can be replaced with a normal far call (or iret). See also http://wiki.osdev.org/Getting_to_Ring_3#The_TSS and http://wiki.osdev.org/Context_Switching

Thanks for your interest. Maybe some day v86 will support task switching, but it's an unused and relatively large feature, so I probably won't work on it in the near future.

**Comment by perlun** [comment]
Thanks for the detailed answer @copy! You're right, we are using hardware multitasking; it is a reasonably simple way to implement multitasking in a kernel (you don't have to take so much responsibility for preserving/restoring register state etc).

So, unfortunately, getting rid of it is _not_ that easy, definitely 1-5 days of work or something. Nonetheless, I've filed a bug about it: https://github.com/chaos4ever/chaos/issues/91

Thanks for your time. I will close this issue for now. Feel free to reopen it if you consider this a "bug"; we could also add it to some list of "unsupported x86 functionality in v86" if you have such a thing.

**Comment by thomasdstewart** [comment]
Hi,

I also ran into this when trying older Linux distros, eg Debian 2.1 (Sink) uses Linux 2.0.38. I get the message in the console: "uncaught exception: Unimplemented: load system segment descriptor, type = 9 (Available 386 TSS)".

I did start to read some of the specs around hardware multitasking and started to have a look at the v86 code but didn't delve too far. So this is mostly just information.

Kind Regards
Tom
